optuna/multi_objective/samplers/_nsga2.py

```diff
@@ -0,0 +1,254 @@
+
+    """
+
+    def __init__(
+        self,
+        seed: Optional[int] = None,
+        population_size: int = 50,
+        mutation_prob: float = 0.1,
+        crossover_prob: float = 0.2,
+    ) -> None:
+        # TODO(ohta): Reconsider the default value of each parameter.
+
+        if not isinstance(population_size, int):
+            raise ValueError("`population_size` must be an integer value.")
+
+        if population_size < 2:
+            raise ValueError("`population_size` must be greater or equal to 2.")
+
+        if not (0.0 <= mutation_prob <= 1.0):
+            raise ValueError("`mutation_prob` must be an float value within the range [0.0, 1.0).")
+
+        if not (0.0 <= crossover_prob <= 1.0):
+            raise ValueError(
+                "`crossover_prob` must be an float value within the range [0.0, 1.0)."
+            )
+
+        self._population_size = population_size
+        self._crossover_prob = crossover_prob
+        self._mutation_prob = mutation_prob
+        self._random_sampler = multi_objective.samplers.RandomMultiObjectiveSampler(seed=seed)
+        self._rng = np.random.RandomState(seed)
+
+    def infer_relative_search_space(
+        self,
+        study: "multi_objective.study.MultiObjectiveStudy",
+        trial: "multi_objective.trial.FrozenMultiObjectiveTrial",
+    ) -> Dict[str, BaseDistribution]:
+        return {}
+
+    def sample_relative(
+        self,
+        study: "multi_objective.study.MultiObjectiveStudy",
+        trial: "multi_objective.trial.FrozenMultiObjectiveTrial",
+        search_space: Dict[str, BaseDistribution],
+    ) -> Dict[str, Any]:
+        parent_generation, population = self._collect_parent_population(study)
+        trial_id = trial._trial_id
+
+        if len(population) == 0:
+            generation = 0
+            study._storage.set_trial_system_attr(trial_id, _GENERATION_KEY, generation)
+        else:
+            p0 = self._select_parent(study, population)
+            p1 = self._select_parent(study, population)
+
+            generation = parent_generation + 1
+            study._storage.set_trial_system_attr(trial_id, _GENERATION_KEY, generation)
+            study._storage.set_trial_system_attr(trial_id, _PARENTS_KEY, [p0.number, p1.number])
+
+        return {}
+
+    def sample_independent(
+        self,
+        study: "multi_objective.study.MultiObjectiveStudy",
+        trial: "multi_objective.trial.FrozenMultiObjectiveTrial",
+        param_name: str,
+        param_distribution: BaseDistribution,
+    ) -> Any:
+        if _PARENTS_KEY not in trial.system_attrs:
+            return self._random_sampler.sample_independent(
+                study, trial, param_name, param_distribution
+            )
+
+        trials = study.get_trials(deepcopy=False)
+        p0, p1 = trial.system_attrs[_PARENTS_KEY]
+
+        param = trials[p0].params.get(param_name, None)
```

Comment:
> I think trials are not necessarily sorted by the trial number. 

The API doc of [Study.get_trials()](https://optuna.readthedocs.io/en/stable/reference/study.html#optuna.study.Study.get_trials) says "The returned trials are ordered by trial number", so I think we can rely on the order (however it seems that we should add the restriction to the storage spec too).

> I suggest using trial_id instead of trial.number and access the trial via get_trial method.
> It's also substantially faster than getting whole trials in most DBs excluding in-memory db.

Thanks. It sounds nice.
I wrote the current code because the current Optuna coding convention recommends accessing trials via `Study` rather than directly using the storage interface as much as possible  (it's not well documented though...). However, your suggestion seems sufficiently reasonable so I'll adopt it.

Reply:
Applied the suggestion: https://github.com/optuna/optuna/pull/1163/commits/cca50c1655af4e92f59901af3a63d98c6d0d9300